### PR TITLE
Adding directory fs

### PR DESCRIPTION
**Pull request by danielstro**

### CODE REVIEW

File: packages/directory/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# @file-services/directory
+<!-- TODO: Remove? replace? -->
```
Comment (danielstro) [inline packages/directory/README.md]: Don't forget

File: packages/directory/src/error-codes.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+export enum FsErrorCodes {
```
Comment (danielstro) [inline packages/directory/src/error-codes.ts]: Do we need this file? Should I put a dedicated error here or just remove it?

File: packages/directory/test/directory-fs-contract.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+        it('Can access a file from a relative path', async () => {
+            const { fs } = testInput
+            const filePath = 'src/index.ts'
+
+            expect((await fs.stat(filePath)).isFile()).to.equal(true)
+            expect(await fs.readFile(filePath)).to.eql(SAMPLE_CONTENT)
+        })
+
+        it('Cannot access a file outside of the base path', async () => {
```
Comment (danielstro) [inline packages/directory/test/directory-fs-contract.ts]: Maybe add a test with a full path that is outside of the base path?

File: packages/utils/README.md [hunk 4]
```diff
@@ -6,3 +6,29 @@ Common file system utility functions.
 # API
 
 - `syncToAsyncFs`: convert a sync-only file system implementation into an async one.
+- `directory-fs`: A file system wrapper that adds scoping to any sync/async, file system implementation.
```
Comment (AviVahl) [inline packages/utils/README.md]:
```suggestion
- `createDirectoryFs`: A file system wrapper that adds directory scoping to any sync/async, base file system implementation.
```

File: packages/utils/README.md [hunk 5]
```diff
@@ -6,3 +6,29 @@ Common file system utility functions.
...
+yarn add @file-services/directory
+```
+
+Then, use the programmatic API:
+```ts
+import { fs } from 'some-fs'
+import { createDirectoryFs } from '@file-services/directory'
+
+const basePath = '/projects/secret-folder'
+const dFs = createDirectoryFs(fs, basePath)
```
Comment (AviVahl) [inline packages/utils/README.md]: directoryFs

File: packages/utils/README.md [hunk 6]
```diff
@@ -6,3 +6,29 @@ Common file system utility functions.
...
 
 - `syncToAsyncFs`: convert a sync-only file system implementation into an async one.
+- `directory-fs`: A file system wrapper that adds scoping to any sync/async, file system implementation.
+
+# Usage
+
+## Directory file system
+Install library in project:
+```sh
+yarn add @file-services/directory
```
Comment (AviVahl) [inline packages/utils/README.md]: /utils

File: packages/utils/README.md [hunk 7]
```diff
@@ -6,3 +6,29 @@ Common file system utility functions.
...
+```
+
+Then, use the programmatic API:
+```ts
+import { fs } from 'some-fs'
+import { createDirectoryFs } from '@file-services/directory'
```
Comment (AviVahl) [inline packages/utils/README.md]: /utils

File: packages/utils/README.md [hunk 8]
```diff
@@ -6,3 +6,29 @@ Common file system utility functions.
...
+
+## Directory file system
+Install library in project:
+```sh
+yarn add @file-services/directory
+```
+
+Then, use the programmatic API:
+```ts
+import { fs } from 'some-fs'
```
Comment (AviVahl) [inline packages/utils/README.md]: might be better to give an example with `@file-services/node` or something

File: packages/utils/README.md [hunk 9]
```diff
@@ -6,3 +6,29 @@ Common file system utility functions.
...
+import { createDirectoryFs } from '@file-services/directory'
+
+const basePath = '/projects/secret-folder'
```
Comment (AviVahl) [inline packages/utils/README.md]: /path/to/some/folder

File: packages/utils/README.md [hunk 10]
```diff
@@ -6,3 +6,29 @@ Common file system utility functions.
 # API
 
 - `syncToAsyncFs`: convert a sync-only file system implementation into an async one.
+- `directory-fs`: A file system wrapper that adds scoping to any sync/async, file system implementation.
+
+# Usage
+
+## Directory file system
```
Comment (AviVahl) [inline packages/utils/README.md]: Directory Scoped File System

File: packages/utils/src/directory-fs.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
+import {IBaseFileSystem, WatchEventListener} from '@file-services/types'
+
+export function createDirectoryFs(fs: IBaseFileSystem, basePath: string): IBaseFileSystem {
+    const joinPath = (path: string) => {
+        const joinedPath = fs.path.join(basePath, path)
+        const relative = fs.path.relative(basePath, joinedPath)
+        if (relative.includes('..')) {
```
Comment (AviVahl) [inline packages/utils/src/directory-fs.ts]: - ```startsWith(`..${path.sep}`)```
- destructure path functions above `joinPath`

File: packages/utils/src/directory-fs.ts [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+            throw new Error(`path ${path} is outside of home directory`)
```
Comment (AviVahl) [inline packages/utils/src/directory-fs.ts]: scoped directory

File: packages/utils/src/directory-fs.ts [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+                    path: fs.path.relative(basePath, e.path)
+                })
+                watchListeners.set(listener, relativePathListener)
+                fs.watchService.addListener(relativePathListener)
+            },
+            removeListener: listener => {
+                const relativePathListener = watchListeners.get(listener)
+                if (relativePathListener) {
+                    fs.watchService.removeListener(relativePathListener)
+                    watchListeners.delete(listener)
```
Comment (AviVahl) [inline packages/utils/src/directory-fs.ts]: shouldn't you delete `relativePathListener` instead?

File: packages/utils/src/index.ts [hunk 14]
```diff
@@ -1,2 +1,3 @@
 export { syncToAsyncFs } from './sync-to-async-fs'
 export { createAsyncFileSystem, createSyncFileSystem } from './create-extended-api'
+export {createDirectoryFs} from './directory-fs'
```
Comment (AviVahl) [inline packages/utils/src/index.ts]: spaces around `createDirectoryFs` for consistency

File: packages/utils/test/directory-fs.spec.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+                        'index.ts': SAMPLE_CONTENT
+                    }
+                },
+                'outside-scope-file.ts': SAMPLE_CONTENT
+            }),
+            basePath
+        )
+    })
+
+    it('can access a file from a relative path', async () => {
```
Comment (AviVahl) [inline packages/utils/test/directory-fs.spec.ts]: using a relative path

**Pull request merged**